### ISSUE TITLE
Gocode completely doesn't work since version golang 1.10

### ISSUE THREAD

**Issue by monoflash** [nsf/gocode#509]
Gocode completely doesn't work since version golang 1.10.
The utility does not return anything after updating golang.
If roll back golang to version 9, then everything starts working again.
:(

**Comment by nsf** [comment]
Might help: https://github.com/nsf/gocode/issues/500

tl;dr: Go has package cache now, with 1.10 you can with hacks make gocode work, but in the long term who knows what happens next. Start looking for alternatives or hope and pray that I'll end up writing next version of the gocode relatively soon (probably it will have a different name).

**Comment by kovetskiy** [comment]
@nsf if you are going to write new version of gocode that will work with source code (instead of package cache), can we cooperate on this task? I don't really want to do it alone and maintain it for years, but it could be easier if several developers would work on that task at the same time.

**Comment by nsf** [comment]
@kovetskiy no, due to unpredictable schedule I don't see how it's possible. I might be coding it every day for a week and then doing zero progress for a few weeks, etc. That's just how open source works. there is nothing to colaborate over until there is a major codebase written and ready. Next gocode will be most likely called "gls" which stands for Go Language Server or Service. It will most likely implement the langauge server protocol by sourcegraph. E.g. similar to: https://github.com/sourcegraph/go-langserver. But having a compatibility mode in mind if possible for existing gocode text editor integrations. What I want is to take full advantage of this protocol and implement everything starting from syntax/semantic checks, goto, "codelense", find usages of, autocompletion, quick fixes, etc. It's a lot of work, but it needs to be done this way and no other way is acceptable. Proper and full source language server protocol implementation.

P.S. If you don't want to maintain it, then don't do it at all.

**Comment by kovetskiy** [comment]
@nsf I mean that I don't want to maintain it if I would be alone as you do with gocode. That's the difference.

**Comment by nsf** [comment]
Well, listen, when and if I have some code to show, then we can talk about collaboration. It's pointless to collaborate over something that doesn't even exist yet. I don't know how it will look, what kind of libs it will use, what it will do exactly, etc. And I don't want to discuss those details either.

**Comment by kovetskiy** [comment]
@nsf ok, no problem, just ping me if I have a chance to help you with any related issue. I will be glad to make a contribution.

**Comment by mdempsky** [comment]
FYI, I just added a -source flag to mdempsky/gocode to make it generate suggestions from source, instead of from compiled package definitions.

Caveat: the main branch of mdempsky/gocode doesn't implement any caching. For compiled imports, this is moderately slower (but IMO still acceptable for interactive use, ~200ms in large packages), but source imports are substantially slower (~1.5s for each completion, even in small packages).

I've prototyped caching in the [cache](https://github.com/mdempsky/gocode/commits/cache) branch. It's currently for compiled packages only (and standard cmd/go workspaces; no gb support), but it brings performance on par with nsf/gocode: ~30ms to ~40ms when cached.

I'd really like to implement this functionality directly into a langserver implementation so that it can make use of langserver's VFS functionality.

**Comment by nsf** [comment]
Added a link to "mdempsky/gocode" to README.md. Hopefully @mdempsky, you don't mind.

**Comment by nsf** [comment]
Although I understand that you don't have a magic fix for "package cache" problem yet. The chance that you'll have one is much higher than on my end. That "-source" flag thing looks interesting.

**Comment by mdempsky** [comment]
@nsf Thank you for the link. I don't mind.

**Comment by localvar** [comment]
@mdempsky , I implement a source caching based on your master branch, at [localvar/gocode](https://github.com/localvar/gocode). I hope I could send a PR to you, but seems too difficult for me to do everything right, though I think it works in most cases.
